### PR TITLE
Support working_dir in from_spark

### PR DESCRIPTION
**Pull request by maddiedawson**
Accept `working_dir` as an argument to `Dataset.from_spark`. Setting a non-NFS working directory for Spark workers to materialize to will improve write performance.

### CODE REVIEW

File: src/datasets/packaged_modules/spark/spark.py [hunk 1]
```diff
@@ -85,6 +87,28 @@ def _info(self):
...
+        import pyspark
+
+        def get_arrow_batch_size(it):
+            for batch in it:
+                yield pa.RecordBatch.from_pydict({"batch_bytes": [batch.nbytes]})
+
+        df_num_rows = self.df.count()
+        # Approximate the size of each row (in Arrow format) by averaging over a 100-row sample.
+        approx_bytes_per_row = (
+            self.df.limit(100)
```
Comment (lu-wang-dl) [inline src/datasets/packaged_modules/spark/spark.py:100]: What happened if the number of rows are less than 100?

File: src/datasets/packaged_modules/spark/spark.py [hunk 2]
```diff
@@ -85,6 +87,28 @@ def _info(self):
     def _split_generators(self, dl_manager: datasets.download.download_manager.DownloadManager):
         return [datasets.SplitGenerator(name=datasets.Split.TRAIN)]
 
+    def _repartition_df_if_needed(self, max_shard_size):
```
Comment (lu-wang-dl) [inline src/datasets/packaged_modules/spark/spark.py:90]: Could we add a test for this function?

File: src/datasets/packaged_modules/spark/spark.py (hunk 1 above)
Comment (maddiedawson) [reply src/datasets/packaged_modules/spark/spark.py]: Fixed

File: src/datasets/packaged_modules/spark/spark.py (hunk 2 above)
Comment (maddiedawson) [reply src/datasets/packaged_modules/spark/spark.py:90]: Done

File: src/datasets/packaged_modules/spark/spark.py [hunk 3]
```diff
@@ -85,6 +87,31 @@ def _info(self):
...
+        import pyspark
```
Comment (lu-wang-dl) [inline src/datasets/packaged_modules/spark/spark.py:91]: Do we need to import pyspark inside the function?

**Review by lu-wang-dl — COMMENTED** [review]
Could we also add the support by setting an environment variable for the working directory?

File: src/datasets/packaged_modules/spark/spark.py (hunk 3 above)
Comment (maddiedawson) [reply src/datasets/packaged_modules/spark/spark.py]: I don't think we need it since we imported in the ctor. Removed

**Comment by maddiedawson** [comment]
Added env var

File: src/datasets/packaged_modules/spark/spark.py (hunk 3 above)
Comment (maddiedawson) [reply src/datasets/packaged_modules/spark/spark.py:138]: Never mind, tests failed with NameError so added it back

**Comment by maddiedawson** [comment]
@lhoestq would you be able to review please? :)

**Review by lu-wang-dl — APPROVED** [review]

File: src/datasets/packaged_modules/spark/spark.py [hunk 4]
```diff
@@ -196,6 +225,11 @@ def write_arrow(it):
                     names=["task_id", "num_examples", "num_bytes"],
                 )
 
+            if working_fpath != fpath:
+                for file in os.listdir(os.path.dirname(working_fpath)):
+                    dest = os.path.join(os.path.dirname(fpath), os.path.basename(file))
+                    os.rename(file, dest)
```
Comment (lhoestq) [inline src/datasets/packaged_modules/spark/spark.py:231]: os.rename doesn't always work depending on the source/target disks, you can use shutil.move instead

**Review by lhoestq — COMMENTED** [review]
Nice ! I think `HF_WORKING_DIR` is not a good env variable name for this case, I guess you can remove it.

File: src/datasets/packaged_modules/spark/spark.py (hunk 4 above)
Comment (maddiedawson) [reply src/datasets/packaged_modules/spark/spark.py]: Thanks, done!

**Comment by maddiedawson** [comment]
I removed the env var

**Review by lhoestq — APPROVED** [review]
LGTM :)

**Comment by maddiedawson** [comment]
Thank you!

**Pull request merged**